File: android/app/src/main/kotlin/com/linkos/core/security/E2EEncryption.kt
package com.linkos.core.security

import java.security.KeyPairGenerator
import java.security.interfaces.ECPublicKey
import java.security.spec.ECGenParameterSpec
import java.security.KeyFactory
import java.security.interfaces.ECPrivateKey
import java.security.spec.X509EncodedKeySpec
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import java.security.MessageDigest
import java.security.SecureRandom

/**
 * End-to-end encryption using ECDH P-256 key exchange and AES-256-GCM.
 *
 * Flow:
 * 1. Each device generates an ephemeral ECDH P-256 key pair
 * 2. Public keys exchanged during pairing
 * 3. Shared secret derived via ECDH
 * 4. Session keys derived via HKDF-SHA256
 * 5. All messages encrypted with AES-256-GCM
 */
object E2EEncryption {

    private const val EC_ALGORITHM = "EC"
    private const val EC_CURVE = "secp256r1"  // P-256
    private const val KEY_AGREEMENT = "ECDH"
    private const val AES_ALGORITHM = "AES/GCM/NoPadding"
    private const val HMAC_ALGORITHM = "HmacSHA256"
    private const val GCM_TAG_LENGTH = 128  // bits
    private const val GCM_NONCE_LENGTH = 12  // bytes
    private const val SESSION_KEY_LENGTH = 32  // bytes (256 bits)

    /**
     * Generate a new ECDH P-256 key pair.
     */
    fun generateKeyPair(): java.security.KeyPair {
        val generator = KeyPairGenerator.getInstance(EC_ALGORITHM)
        generator.initialize(ECGenParameterSpec(EC_CURVE))
        return generator.generateKeyPair()
    }

    /**
     * Derive a shared session key from our private key and peer's public key.
     * Uses ECDH + HKDF-SHA256.
     */
    fun deriveSessionKey(
        ownPrivateKey: ECPrivateKey,
        peerPublicKey: ECPublicKey,
        salt: ByteArray = ByteArray(0),
        info: ByteArray = "LinkOS-Session-v1".toByteArray()
    ): ByteArray {
        // ECDH key agreement
        val agreement = KeyAgreement.getInstance(KEY_AGREEMENT)
        agreement.init(ownPrivateKey)
        agreement.doPhase(peerPublicKey, true)
        val sharedSecret = agreement.generateSecret()

        // HKDF extract + expand
        return hkdf(sharedSecret, salt, info, SESSION_KEY_LENGTH)
    }

    /**
     * Encrypt data using AES-256-GCM.
     * Returns: nonce (12 bytes) || ciphertext || tag (16 bytes)
     */
    fun encrypt(plaintext: ByteArray, key: ByteArray): ByteArray {
        val nonce = ByteArray(GCM_NONCE_LENGTH).also { SecureRandom().nextBytes(it) }
        val cipher = Cipher.getInstance(AES_ALGORITHM)
        val spec = GCMParameterSpec(GCM_TAG_LENGTH, nonce)
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), spec)
        val ciphertext = cipher.doFinal(plaintext)
        return nonce + ciphertext
    }

    /**
     * Decrypt data using AES-256-GCM.
     * Input: nonce (12 bytes) || ciphertext || tag (16 bytes)
     */
    fun decrypt(data: ByteArray, key: ByteArray): ByteArray {
        val nonce = data.copyOfRange(0, GCM_NONCE_LENGTH)
        val ciphertext = data.copyOfRange(GCM_NONCE_LENGTH, data.size)
        val cipher = Cipher.getInstance(AES_ALGORITHM)
        val spec = GCMParameterSpec(GCM_TAG_LENGTH, nonce)
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), spec)
        return cipher.doFinal(ciphertext)
    }

    /**
     * Compute HMAC-SHA256 for message integrity.
     */
    fun hmac(data: ByteArray, key: ByteArray): ByteArray {
        val mac = Mac.getInstance(HMAC_ALGORITHM)
        mac.init(SecretKeySpec(key, HMAC_ALGORITHM))
        return mac.doFinal(data)
    }

    /**
     * Verify HMAC-SHA256.
     */
    fun verifyHmac(expectedMac: ByteArray, data: ByteArray, key: ByteArray): Boolean {
        val computed = hmac(data, key)
        return MessageDigest.isEqual(computed, expectedMac)
    }

    /**
     * SHA-256 fingerprint of a public key for display verification.
     */
    fun fingerprint(publicKey: ECPublicKey): String {
        val digest = MessageDigest.getInstance("SHA-256")
        val hash = digest.digest(publicKey.encoded)
        return hash.take(8).joinToString(":") { "%02X".format(it) }
    }

    /**
     * Serialize public key for network transmission.
     */
    fun serializePublicKey(key: ECPublicKey): ByteArray = key.encoded

    /**
     * Deserialize public key from network data.
     */
    fun deserializePublicKey(data: ByteArray): ECPublicKey {
        val keyFactory = KeyFactory.getInstance(EC_ALGORITHM)
        return keyFactory.generatePublic(X509EncodedKeySpec(data)) as ECPublicKey
    }

    // MARK: - HKDF

    private fun hkdf(
        inputKeyMaterial: ByteArray,
        salt: ByteArray,
        info: ByteArray,
        outputLength: Int
    ): ByteArray {
        // Extract
        val prk = hmac(inputKeyMaterial, if (salt.isEmpty()) ByteArray(32) else salt)

        // Expand
        var output = ByteArray(0)
        var t = ByteArray(0)
        var counter: Byte = 1

        while (output.size < outputLength) {
            t = hmac(t + info + byteArrayOf(counter), prk)
            output += t
            counter++
        }

        return output.copyOfRange(0, outputLength)
    }
}

/**
 * Manages an active encrypted session with a peer device.
 */
class EncryptedSession(
    val deviceId: String,
    private val sessionKey: ByteArray,
) {
    val createdAt: Long = System.currentTimeMillis()
    var lastActivity: Long = System.currentTimeMillis()
        private set

    fun encrypt(data: ByteArray): ByteArray {
        lastActivity = System.currentTimeMillis()
        return E2EEncryption.encrypt(data, sessionKey)
    }

    fun decrypt(data: ByteArray): ByteArray {
        lastActivity = System.currentTimeMillis()
        return E2EEncryption.decrypt(data, sessionKey)
    }

    fun sign(data: ByteArray): ByteArray {
        lastActivity = System.currentTimeMillis()
        return E2EEncryption.hmac(data, sessionKey)
    }

    fun verify(signature: ByteArray, data: ByteArray): Boolean =
        E2EEncryption.verifyHmac(signature, data, sessionKey)
}

File: macOS/LinkOS/Core/Security/E2EEncryption.swift
import Foundation
import CryptoKit

/// End-to-end encryption using ECDH P-256 key exchange and AES-256-GCM.
///
/// Flow:
/// 1. Each device generates an ephemeral ECDH P-256 key pair
/// 2. Devices exchange public keys during pairing
/// 3. Shared secret derived via ECDH
/// 4. Session keys derived via HKDF-SHA256
/// 5. All messages encrypted with AES-256-GCM
final class E2EEncryption {
    
    /// Generate a new ECDH P-256 key pair for key exchange.
    static func generateKeyPair() -> (privateKey: P256.KeyAgreement.PrivateKey,
                                       publicKey: P256.KeyAgreement.PublicKey) {
        let privateKey = P256.KeyAgreement.PrivateKey()
        return (privateKey, privateKey.publicKey)
    }
    
    /// Derive a shared secret from our private key and the peer's public key.
    /// Returns a 256-bit symmetric key suitable for AES-256-GCM.
    static func deriveSessionKey(
        ownPrivateKey: P256.KeyAgreement.PrivateKey,
        peerPublicKey: P256.KeyAgreement.PublicKey,
        salt: Data? = nil,
        info: Data = "LinkOS-Session-v1".data(using: .utf8)!
    ) throws -> SymmetricKey {
        let sharedSecret = try ownPrivateKey.sharedSecretFromKeyAgreement(with: peerPublicKey)
        let derivedKey = sharedSecret.hkdfDerivedSymmetricKey(
            using: SHA256.self,
            salt: salt ?? Data(),
            sharedInfo: info,
            outputByteCount: 32  // 256-bit key
        )
        return derivedKey
    }
    
    /// Encrypt data using AES-256-GCM.
    /// Returns: nonce (12 bytes) + ciphertext + tag (16 bytes)
    static func encrypt(_ plaintext: Data, using key: SymmetricKey) throws -> Data {
        let sealedBox = try AES.GCM.seal(plaintext, using: key)
        guard let combined = sealedBox.combined else {
            throw EncryptionError.sealFailed
        }
        return combined
    }
    
    /// Decrypt data using AES-256-GCM.
    /// Input: nonce (12 bytes) + ciphertext + tag (16 bytes)
    static func decrypt(_ ciphertext: Data, using key: SymmetricKey) throws -> Data {
        let sealedBox = try AES.GCM.SealedBox(combined: ciphertext)
        return try AES.GCM.open(sealedBox, using: key)
    }
    
    /// Compute HMAC-SHA256 for message integrity.
    static func hmac(for data: Data, using key: SymmetricKey) -> Data {
        let authCode = HMAC<SHA256>.authenticationCode(for: data, using: key)
        return Data(authCode)
    }
    
    /// Verify HMAC-SHA256 for message integrity.
    static func verifyHMAC(_ mac: Data, for data: Data, using key: SymmetricKey) -> Bool {
        HMAC<SHA256>.isValidAuthenticationCode(mac, authenticating: data, using: key)
    }
    
    /// Generate SHA-256 fingerprint of a public key (for device verification display).
    static func fingerprint(of publicKey: P256.KeyAgreement.PublicKey) -> String {
        let hash = SHA256.hash(data: publicKey.rawRepresentation)
        return hash.prefix(8).map { String(format: "%02X", $0) }.joined(separator: ":")
    }
    
    /// Serialize a public key for network transmission.
    static func serializePublicKey(_ key: P256.KeyAgreement.PublicKey) -> Data {
        key.derRepresentation
    }
    
    /// Deserialize a public key from network data.
    static func deserializePublicKey(_ data: Data) throws -> P256.KeyAgreement.PublicKey {
        try P256.KeyAgreement.PublicKey(derRepresentation: data)
    }
}

// MARK: - Errors

enum EncryptionError: LocalizedError {
    case sealFailed
    case invalidPublicKey
    case keyDerivationFailed
    
    var errorDescription: String? {
        switch self {
        case .sealFailed: return "AES-GCM seal failed"
        case .invalidPublicKey: return "Invalid public key data"
        case .keyDerivationFailed: return "HKDF key derivation failed"
        }
    }
}

// MARK: - Session

/// Manages an active encrypted session with a peer device.
final class EncryptedSession {
    let deviceId: String
    let sessionKey: SymmetricKey
    let createdAt: Date
    private(set) var lastActivity: Date
    
    init(deviceId: String, sessionKey: SymmetricKey) {
        self.deviceId = deviceId
        self.sessionKey = sessionKey
        self.createdAt = Date()
        self.lastActivity = Date()
    }
    
    func encrypt(_ data: Data) throws -> Data {
        lastActivity = Date()
        return try E2EEncryption.encrypt(data, using: sessionKey)
    }
    
    func decrypt(_ data: Data) throws -> Data {
        lastActivity = Date()
        return try E2EEncryption.decrypt(data, using: sessionKey)
    }
    
    func sign(_ data: Data) -> Data {
        lastActivity = Date()
        return E2EEncryption.hmac(for: data, using: sessionKey)
    }
    
    func verify(signature: Data, for data: Data) -> Bool {
        E2EEncryption.verifyHMAC(signature, for: data, using: sessionKey)
    }
}
